### PR TITLE
Make optional setter covariant

### PR DESCRIPTION
**Pull request by JacekLach**
### Before this PR

```
// imagine this comes from a function and you don't know if it's present or absent
Optional<String> specificValue = Optional.of("foo");
RequestArgs args = RequestArgs.builder()
  // optionalArgument is `optional<any>` because different requests take different types
  .optionalArgument(specificValue)
  .build();
args.getOptionalArgument() // returns Optional.of(Optional.of("foo"))
```

### After this PR

```
Optional<String> specificValue = Optional.of("foo");
RequestArgs args = RequestArgs.builder()
  // optionalArgument is `optional<any>` because different requests take different types
  .optionalArgument(specificValue)
  .build();
args.getOptionalArgument() // returns Optional.of("foo")
```

### CODE REVIEW

**Comment by JacekLach** [comment]
Oh, I need to handle primitives specially here. Will fix later

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java [hunk 1]
```diff
@@ -217,6 +224,14 @@ private static CodeBlock typeAwareAssignment(EnrichedField enriched, Type type,
                             ByteBuffer.class)
                     .addStatement("this.$1N.rewind()", spec.name)
                     .build();
+        } else if (type.accept(TypeVisitor.IS_OPTIONAL)) {
```
Comment (markelliot) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java:321]: Why is this change required?

Comment (JacekLach) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java:321]: You can't assign a covariant type to an invariant type, like `Optional<? extends Object> bar = getBar(); Optional<Object> foo = bar;`. You need to either explicitly recast and suppress compiler warnings, or transform the generic type with `Optional.<Object>of(bar.get())`

(actually, I checked and it looks like `Optional<Object> foo = (Optional<Object>) bar;` works without java warnings so maybe that's better)
(oh, actually, it does produce a warning but it seems to be suppressed in generated code by default)

Comment (markelliot) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java:321]: I see. And you didn't want to widen the type everywhere?

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java [hunk 2]
```diff
@@ -217,6 +224,14 @@ private static CodeBlock typeAwareAssignment(EnrichedField enriched, Type type,
...
+            Type innerType = type.accept(TypeVisitor.OPTIONAL).getItemType();
+            CodeBlock nullCheckedValue = Expressions.requireNonNull(
+                    spec.name, enriched.fieldName().get() + " cannot be null");
+            return CodeBlock.builder()
+                    .addStatement("this.$1N = $2L.flatMap($3T::<$4T>::of)",
```
Comment (markelliot) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java]: just filling in the syntax to make it clearer:

```java
foo.flatMap(Optional::<Object>::of)
```

doesn't seem valid to me/locally it does not compile for me.

Comment (markelliot) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java]: there needs to be some testing around both the code gen and the logic of the codegen in this PR

**Comment by uschi2000** [comment]
what's a concrete use case for this? typically, our conjure objects have concrete types and there's no class hierarchy.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java (hunk 2 above)
Comment (JacekLach) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java]: Yeah, the extra `::` was a typo that I already had fixed locally but forgot to push. I added a minor test but currently the infra for adding new cases is missing; will come back to it after https://github.com/palantir/conjure-java/issues/10 is fixed

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java (hunk 1 above)
Comment (JacekLach) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java:321]: We could widen the type for optionals everywhere, but that feels a bigger / more scary change to me. If maintainers think that's a better approach, I'm happy to follow suit.

Intuitively I feel like it might require changes in many places; if `ConjureOptional.getItem()` starts returning `Optional<? extends FOO>` then a lot of code that currently expects `Optional<FOO>` will break; and I don't see an advantage to getting a wildcard type when reading

**Comment by JacekLach** [comment]
@uschi2000 the biggest benefit is for `Optional<any>`, which sees some usage.

With current bindings, you get this gotcha:

```
 // imagine this comes from a function and you don't know if it's present or absent
Optional<String> specificValue = Optional.of("foo");
RequestArgs args = RequestArgs.builder()
  // optionalArgument is `optional<any>` because different requests take different types
  .optionalArgument(specificValue)
  .build();
args.getOptionalArgument() // returns Optional.of(Optional.of("foo"))
```

With wildcard builder you get `Optional.of("foo")` in the last example, which I think you more often expect. Of course it means if you actually want a directly nested optional in an untyped field you need to put in some more boilerplate; but that seems a very rare occurrence.

File: conjure-java-core/src/test/java/com/palantir/conjure/java/CovariantsTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.junit.Test;
+
+public class CovariantsTest {
+
+    @Test
+    public void covariantOptionalBuilderTest() {
+        String value = "foo";
+        Optional<String> maybeValue = Optional.of(value);
+
+        // FIXME: this doesn't compile right now because changes to example-types.yml aren't picked up
```
Comment (uschi2000) [inline conjure-java-core/src/test/java/com/palantir/conjure/java/CovariantsTest.java]: there's a test here that re-generates the code when you comment out a `if (REGEN)` - ish statement

Comment (JacekLach) [reply conjure-java-core/src/test/java/com/palantir/conjure/java/CovariantsTest.java]: I believe that regenerates the java code based on the `.json` definitions, but there's nothing that regenerates the `.json` definitions right now, see #10

**Comment by JacekLach** [comment]
Green!

**Comment by markelliot** [comment]
Given the use-case, I wonder if we only want to be covariant for `any` to limit abuse.

**Comment by JacekLach** [comment]
I think it's also possibly handy for things like collections (being able to put an `Optional<SortedSet>` into a `optional<set<string>>`), but it is only in the case of `any` that without this change you transparently use the wrong overload, instead of getting a compilation error. So I'm OK with just doing this for `any` at first, and extending if real use-case shows up.

**Comment by JacekLach** [comment]
Another related annoyance is the static builders for collections:

```
    public static CovariantListExample of(List<Object> items) {
        return builder().items(items).build();
    }
```

but at a quick glance it doesn't seem trivial to widen the parameters there (the `BeanBuilderGenerator` and `BeanGenerator` seem to operate differently)

File: conjure-java-core/src/integrationInput/java/com/palantir/product/OptionalExample.java [hunk 4]
```diff
@@ -99,7 +99,7 @@ public Builder from(OptionalExample other) {
 
         @JsonSetter("item")
         public Builder item(Optional<String> item) {
-            this.item = Objects.requireNonNull(item, "item cannot be null");
+            this.item = (Optional<String>) Objects.requireNonNull(item, "item cannot be null");
```
Comment (dansanduleac) [inline conjure-java-core/src/integrationInput/java/com/palantir/product/OptionalExample.java]: this cast doesn't seem necessary

Comment (JacekLach) [reply conjure-java-core/src/integrationInput/java/com/palantir/product/OptionalExample.java]: yep, missed out on fixing tests after 5435159

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java (hunk 1 above)
Comment (dansanduleac) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java:321]: I don't think there's much value in _returning_ `Optional<? extends FOO>`, means you now have to propagate `?` everywhere, and you can't assign it to a variable of type `Optional<FOO>` without a cast.
The value is there only when it's a parameter, since the compiler knows that `FOO` or e.g. `BAR <: FOO` both satisfy `? extends FOO`.

File: conjure-java-core/src/integrationInput/java/com/palantir/product/OptionalExample.java (hunk 4 above)
Comment (JacekLach) [reply conjure-java-core/src/integrationInput/java/com/palantir/product/OptionalExample.java]: 530b826 should address this and fix tests

File: conjure-java-core/src/integrationInput/java/com/palantir/product/ListExample.java [hunk 5]
```diff
@@ -179,15 +179,15 @@ public Builder primitiveItems(int primitiveItems) {
         }
 
         @JsonSetter("doubleItems")
-        public Builder doubleItems(Iterable<Double> doubleItems) {
+        public Builder doubleItems(Iterable<? extends Double> doubleItems) {
```
Comment (iamdanfox) [inline conjure-java-core/src/integrationInput/java/com/palantir/product/ListExample.java]: As a rule of thumb, we've tried to make conjure-java generate code that is as close as possible to what a human might write.  Can we avoid the `? extends` for known classes like `String`, `Double`, `Integer` etc?

File: conjure-java-core/src/main/java/com/palantir/conjure/spec/ConjureDefinition.java [hunk 6]
```diff
@@ -158,14 +158,14 @@ public Builder version(int version) {
         }
 
         @JsonSetter("errors")
-        public Builder errors(Iterable<ErrorDefinition> errors) {
+        public Builder errors(Iterable<? extends ErrorDefinition> errors) {
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/spec/ConjureDefinition.java]: I'm also not a fan of this. When a builder accepts a conjure type, we know it is final so there's no point adding `? extends`

**Comment by iamdanfox** [comment]
@JacekLach Dan and I had a couple of CR comments above - what do you think?

**Comment by JacekLach** [comment]
I agree with the sentiment behind them, just didn't have time to look into how to make it happen. The answer is probably to inspect if the type is final somehow.

**Comment by JacekLach** [comment]
@iamdanfox I think this does it.

I'll let the `.of` constructors be for now even though they make me sad, since at least you can now use the builder when you need to :P

File: conjure-java-core/src/main/java/com/palantir/conjure/spec/EndpointDefinition.java [hunk 7]
```diff
@@ -290,14 +290,14 @@ public Builder auth(AuthType auth) {
...
-        public Builder args(Iterable<ArgumentDefinition> args) {
+        public Builder args(Iterable<? extends ArgumentDefinition> args) {
             this.args.clear();
             ConjureCollections.addAll(
                     this.args, Objects.requireNonNull(args, "args cannot be null"));
             return this;
         }
 
-        public Builder addAllArgs(Iterable<ArgumentDefinition> args) {
+        public Builder addAllArgs(Iterable<? extends ArgumentDefinition> args) {
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/spec/EndpointDefinition.java]: I think these shouldn't have changed?

**Review by iamdanfox — APPROVED** [review]
I think this makes sense to me as it prevents the nasty footgun you encountered (putting `Optional.of(x)` into a builder method, resulting in an unwanted double-nested optional).

The code necessary to achieve this does seem pretty hard to read, but I think the outcome is probably worth it.
